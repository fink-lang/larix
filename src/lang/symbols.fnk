{symbol} = import '@fink/prattler/symbols'
{curr_loc} = import '@fink/prattler'

{next_is_new_expr} = import './block/indentation'
{single_expression} = import './block/expr'



infix_led = fn type, op, lbp: fn ctx, left:
  {loc: {start}} = left
  [right, next_ctx] = single_expression:: ctx, lbp
  {loc: {end}} = right

  [{type, op, left, right, loc: {start, end}}, next_ctx]


left_binding = fn lbp: fn ctx:
  match ctx:
    next_is_new_expr:: ?:
      0
    else:
      lbp


non_bining = fn: fn: 0


infix = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp + 1
  }


infix_right = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp - 1
  }


# TODO: why is this not type, op?
prefix = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn lbp: fn ctx:
      {start} = curr_loc:: ctx
      [right, next_ctx] = single_expression:: ctx, lbp
      {end} = right.loc

      [{type, op, right, loc: {start, end}}, next_ctx]
  }


literal = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn: fn ctx:
      loc = curr_loc:: ctx
      # value = curr_value:: ctx
      [{type, value: op, loc}, ctx]
  }


terminator = fn op:
  {
    ...symbol:: op

    lbp: left_binding
  }
