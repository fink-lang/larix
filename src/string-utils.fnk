unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
