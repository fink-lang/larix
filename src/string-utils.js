import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

