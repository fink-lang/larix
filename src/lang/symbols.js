import {symbol as parser_symbol} from '@fink/prattler/symbols';
import {next_loc} from '@fink/prattler';
import {indentation} from './indentation';


export const symbol = (id)=> ({
  ...parser_symbol(id),

  lbp: (lbp)=> (ctx)=> {
    if (next_loc(ctx).start.column <= indentation(ctx)) {
      return 0;
    }
    return lbp;
  }
});
