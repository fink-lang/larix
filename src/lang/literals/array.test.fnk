{decribe, it, expect, to_match_snapshot, to_throw} = import '../../testing/jest'

{parse_expr} = import '../../'


describe:: 'array [...]', fn:

  it:: 'parses empty: []', fn:
    expect::
      parse_expr:: `[]`
      to_match_snapshot


  it:: 'parses single elemement: [1]', fn:
    expect::
      parse_expr:: `[1]`
      to_match_snapshot


  it:: 'parses multiple elements: [1, 2]', fn:
    expect::
      parse_expr:: `[1, 2]`
      to_match_snapshot

    expect::
      parse_expr:: `[1, 2, 3]`
      to_match_snapshot


  it:: 'parses leading commas: [,, foo]', fn:
    expect::
      parse_expr:: `[,, foo]`
      to_match_snapshot


  it:: 'parses trailing commas: [...foo,,]', fn:
    expect::
      parse_expr:: `[...foo, ]`
      to_match_snapshot

    expect::
      parse_expr:: `[...foo, , ]`
      to_match_snapshot

    expect::
      parse_expr:: `[, ...foo, ]`
      to_match_snapshot


describe:: 'array block', fn:
  it:: 'parses empty', fn:
    expect::
      parse_expr:: `
        [
        ]`
      to_match_snapshot


  it:: 'parses single elemement', fn:
    expect::
      parse_expr:: `
        [
          1
        ]`
      to_match_snapshot


  it:: 'parses multiple elements', fn:
    expect::
      parse_expr:: `
        [
          1
          2
        ]`
      to_match_snapshot

    expect::
      parse_expr:: `
        [
          1, 2
        ]`
      to_match_snapshot


  it:: 'parses leading commas', fn:
    expect::
      parse_expr:: `
        [
          ,
          ,
          foo
        ]`
      to_match_snapshot

    expect::
      parse_expr:: `
        [
          ,,
          foo
        ]`
      to_match_snapshot


  it:: 'parses trailing commas', fn:
    expect::
      parse_expr:: `
        [
          foo
          ,
        ]`
      to_match_snapshot

    expect::
      parse_expr:: `
        [
          foo
          ,
          ,
        ]`
      to_match_snapshot

    expect::
      parse_expr:: `
        [
          ,
          ...foo
          ,
        ]`
      to_match_snapshot


describe:: 'list', fn:
  it:: 'parses single elemement', fn:
    expect::
      parse_expr:: `
        list:
          1
        `
      to_match_snapshot


  it:: 'parses multiple elements', fn:
    expect::
      parse_expr:: `
        list:
          1
          2
        `
      to_match_snapshot


  it:: 'parses leading commas', fn:
    expect::
      parse_expr:: `
        list:
          ,
          ,
          foo
        `
      to_match_snapshot

    expect::
      parse_expr:: `
        list:
          ,,
          foo
        `
      to_match_snapshot


  it:: 'parses trailing commas', fn:
    expect::
      parse_expr:: `
        list:
          foo
          ,
        `
      to_match_snapshot

    expect::
      parse_expr:: `
        list:
          foo
          ,
          ,
        `
      to_match_snapshot

    expect::
      parse_expr:: `
        list:
          ,
          ...foo
          ,
        `
      to_match_snapshot


describe:: 'array [...] - parsing failures', fn:
  it:: `throws when missing ']'`, fn:
    expect::
      fn: parse_expr:: `[`, 'test.fnk'

      to_throw:: `
        test.fnk:1:1
        1| [
            ^

        Expected ',' or indented new line or ']'.`


  it:: `throws when missing ']'`, fn:
    expect::
      fn: parse_expr:: `[1)]`, 'test.fnk'

      to_throw:: `
        test.fnk:1:2
        1| [1)]
             ^

        Expected ',' or new-line or end of expressions.`


