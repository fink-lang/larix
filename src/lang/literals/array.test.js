import {other_token} from '@fink/prattler/symbols';

import {parse_expr} from '../../';


import {strip_block} from '../../string-utils';


describe('array [...]', ()=> {

  it('parses empty: []', ()=> {
    expect(
      parse_expr(`[]`)
    ).toEqual({
      type: 'array',
      exprs: [],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 2, line: 1, column: 2}
      }
    });
  });


  it('parses single elemement: [1]', ()=> {
    expect(
      parse_expr(`[1]`)
    ).toEqual({
      type: 'array',
      exprs: [parse_expr(` 1`)],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 3, line: 1, column: 3}
      }
    });
  });


  it('parses multiple elements: [1, 2]', ()=> {
    expect(
      parse_expr(`[1, 2]`)
    ).toEqual({
      type: 'array',
      exprs: [
        parse_expr(` 1`),
        parse_expr(`    2`)
      ],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 6, line: 1, column: 6}
      }
    });
  });

  // TODO: should that be supported?
  it('parses leading commas: [,, foo]', ()=> {
    expect(
      parse_expr(`[,, foo]`)
    ).toEqual({
      type: 'array',
      exprs: [
        null,
        null,
        parse_expr(`    foo`)
      ],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 8, line: 1, column: 8}
      }
    });
  });


  // TODO: should that be supported?
  it('parses dangling comma: [1, 2,]', ()=> {
    expect(
      parse_expr(`[1, 2,]`)
    ).toEqual({
      type: 'array',
      exprs: [
        parse_expr(` 1`),
        parse_expr(`    2`)
      ],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 7, line: 1, column: 7}
      }
    });
  });
});


describe('array [...] - parsing failures', ()=> {
  it('throws when missing `]`', ()=> {
    expect(
      ()=> parse_expr(`[`)
    ).toThrow(strip_block`
      Expected ']' but found Symbol(end):
      1| [
          ^`
    );
  });

  it('throws when missing `,`', ()=> {
    expect(
      ()=> parse_expr(`[1)]`)
    ).toThrow(strip_block`
      Expected ',' but found ')':
      1| [1)]
           ^`
    );
  });
});

