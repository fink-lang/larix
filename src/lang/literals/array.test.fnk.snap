// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array [...] parses empty: [] 1`] = `
array (1:0-1:2)

`;

exports[`array [...] parses leading commas: [,, foo] 1`] = `
array (1:0-1:8)
  empty (1:2-1:2)
  empty (1:3-1:3)
  ident (1:4-1:7) foo
`;

exports[`array [...] parses multiple elements: [1, 2] 1`] = `
array (1:0-1:6)
  number (1:1-1:2) 1
  number (1:4-1:5) 2
`;

exports[`array [...] parses multiple elements: [1, 2] 2`] = `
array (1:0-1:9)
  number (1:1-1:2) 1
  number (1:4-1:5) 2
  number (1:7-1:8) 3
`;

exports[`array [...] parses single elemement: [1] 1`] = `
array (1:0-1:3)
  number (1:1-1:2) 1
`;

exports[`array [...] parses trailing commas: [...foo,,] 1`] = `
array (1:0-1:10)
  spread ... (1:1-1:7)
    ident (1:4-1:7) foo
  empty (1:10-1:10)
`;

exports[`array [...] parses trailing commas: [...foo,,] 2`] = `
array (1:0-1:12)
  spread ... (1:1-1:7)
    ident (1:4-1:7) foo
  empty (1:10-1:10)
  empty (1:12-1:12)
`;

exports[`array [...] parses trailing commas: [...foo,,] 3`] = `
array (1:0-1:12)
  empty (1:2-1:2)
  spread ... (1:3-1:9)
    ident (1:6-1:9) foo
  empty (1:12-1:12)
`;

exports[`array block parses empty 1`] = `
array (1:0-2:1)

`;

exports[`array block parses leading commas 1`] = `
array (1:0-5:1)
  empty (2:3-2:3)
  empty (3:3-3:3)
  ident (4:2-4:5) foo
`;

exports[`array block parses leading commas 2`] = `
array (1:0-4:1)
  empty (2:3-2:3)
  empty (2:4-2:4)
  ident (3:2-3:5) foo
`;

exports[`array block parses multiple elements 1`] = `
array (1:0-4:1)
  number (2:2-2:3) 1
  number (3:2-3:3) 2
`;

exports[`array block parses multiple elements 2`] = `
array (1:0-3:1)
  number (2:2-2:3) 1
  number (2:5-2:6) 2
`;

exports[`array block parses single elemement 1`] = `
array (1:0-3:1)
  number (2:2-2:3) 1
`;

exports[`array block parses trailing commas 1`] = `
array (1:0-4:1)
  ident (2:2-2:5) foo
  empty (3:3-3:3)
`;

exports[`array block parses trailing commas 2`] = `
array (1:0-5:1)
  ident (2:2-2:5) foo
  empty (3:3-3:3)
  empty (4:3-4:3)
`;

exports[`array block parses trailing commas 3`] = `
array (1:0-5:1)
  empty (2:3-2:3)
  spread ... (3:2-3:8)
    ident (3:5-3:8) foo
  empty (4:3-4:3)
`;

exports[`list parses leading commas 1`] = `
array list (1:0-5:0)

  :
  empty (2:3-2:3)
  empty (3:3-3:3)
  ident (4:2-4:5) foo
`;

exports[`list parses leading commas 2`] = `
array list (1:0-4:0)

  :
  empty (2:3-2:3)
  empty (2:4-2:4)
  ident (3:2-3:5) foo
`;

exports[`list parses multiple elements 1`] = `
array list (1:0-4:0)

  :
  number (2:2-2:3) 1
  number (3:2-3:3) 2
`;

exports[`list parses single elemement 1`] = `
array list (1:0-3:0)

  :
  number (2:2-2:3) 1
`;

exports[`list parses trailing commas 1`] = `
array list (1:0-4:0)

  :
  ident (2:2-2:5) foo
  empty (3:3-3:3)
`;

exports[`list parses trailing commas 2`] = `
array list (1:0-5:0)

  :
  ident (2:2-2:5) foo
  empty (3:3-3:3)
  empty (4:3-4:3)
`;

exports[`list parses trailing commas 3`] = `
array list (1:0-5:0)

  :
  empty (2:3-2:3)
  spread ... (3:2-3:8)
    ident (3:5-3:8) foo
  empty (4:3-4:3)
`;
